### PR TITLE
Fix msgn

### PR DESCRIPTION
**Pull request by g8sqh**
## What does this PR do?
Allows MSG native format files to be read in and geo-referenced reasonably accurately
Works (in qgis) for all three MSG satellites, and HRV and normal resolution

TODO  - check on multiple image sets to optimize accuracy. Consider whether refraction at the image edge may become significant

## What are related issues/pull requests?
Issue #5866 

## Tasklist

 - [ ] ADD YOUR TASKS HERE
 - [ ] Add test case(s)
 - [ ] Add documentation
 - [ ] Updated Python API documentation (swig/include/python/docs/)
 - [ ] Review
 - [ ] Adjust for comments
 - [ ] All CI builds and checks have passed

## Environment

Provide environment details, if relevant:

* OS:
* Compiler:

### CODE REVIEW

**Comment by g8sqh** [comment]
As requested - a single commit that (I hope) preserves the existing layout

File: frmts/msgn/msgndataset.cpp [hunk 1]
```diff
@@ -522,41 +601,82 @@ GDALDataset *MSGNDataset::Open( GDALOpenInfo * poOpenInfo )
...
     poDS->adfGeoTransform[5] = -pixel_gsd_y;
 
     OGRSpatialReference oSRS;
 
+    //int i;
+    for (i=0; i<6; ++i)
+    {
+        CPLDebugOnly("MSGN", "adf transform %12.3f\n", poDS->adfGeoTransform[i]);
+    }
+
```
Comment (rouault) [inline frmts/msgn/msgndataset.cpp:656]:
```suggestion
```

File: frmts/msgn/msgndataset.cpp [hunk 2]
```diff
@@ -522,41 +601,82 @@ GDALDataset *MSGNDataset::Open( GDALOpenInfo * poOpenInfo )
...
+               // false northing to handle the fact RSS is only 1/3 disk
+               pixel_gsd_y *((poDS->m_Shape == RSS) ? ((open_mode != MODE_HRV)  ?
+                                   -(idr.plannedCoverage_visir.southernLinePlanned -1) : // MSG-3 vis/NIR N-S P2 
+                                   -(idr.plannedCoverage_hrv.lowerSouthLinePlanned +1)) : // MSG-3 HRV N-S P2 -ve N
+                     0.0 ));
 
     CPLFree(poDS->pszProjection);
     poDS->pszProjection = nullptr;
     oSRS.exportToWkt( &(poDS->pszProjection) );
+    CPLDebug("MSGN", "WKT: %s", poDS->pszProjection);
```
Comment (rouault) [inline frmts/msgn/msgndataset.cpp:678]:
```suggestion
```

File: frmts/msgn/msgndataset.cpp [hunk 3]
```diff
@@ -472,18 +491,78 @@ GDALDataset *MSGNDataset::Open( GDALOpenInfo * poOpenInfo )
...
+                 // full expected width
+                 idr.plannedCoverage_hrv.lowerWestColumnPlanned == idr.plannedCoverage_hrv.lowerEastColumnPlanned + nRawHRVColumns - 1 &&
+                 idr.plannedCoverage_hrv.lowerSouthLinePlanned > 1 &&
+                 idr.plannedCoverage_hrv.lowerSouthLinePlanned < idr.referencegrid_hrv.numberOfLines &&
+                 idr.plannedCoverage_hrv.lowerEastColumnPlanned >= 1 &&
+                 idr.plannedCoverage_hrv.lowerWestColumnPlanned <= poDS->nRasterXSize * 3 &&
+                 // full height
+                 idr.plannedCoverage_hrv.lowerNorthLinePlanned == idr.plannedCoverage_hrv.lowerSouthLinePlanned + poDS->nRasterYSize -1
+                 )
+    {
```
Comment (rouault) [inline frmts/msgn/msgndataset.cpp:513]:
```suggestion
        {
```

**Comment by rouault** [comment]
The backport to `release/3.5` failed:
```
The process '/usr/bin/git' failed with exit code 128
```
<details>
<summary>stderr</summary>

```
error: commit aa3e1e39f52228140fdf1d4c64c63ca10d84e641 is a merge but no -m option was given.
fatal: cherry-pick failed

```
</details>

<details>
<summary>stdout</summary>

```
Auto-merging frmts/msgn/msgndataset.cpp
[backport-6244-to-release/3.5 50ef97e169] Should now handle vis/NIR and HRV for each of MSG4, MSG3(RSS) and MSG2(IODC) with reasonable accuracy in reprojection
 Author: David Hutchinson <djch-github@hutchhome.co.uk>
 Date: Sun Aug 21 23:06:05 2022 +0100
 5 files changed, 398 insertions(+), 70 deletions(-)

```
</details>

To backport manually, run these commands in your terminal:
```bash
# Fetch latest updates from GitHub
git fetch
# Create a new working tree
git worktree add .worktrees/backport-release/3.5 release/3.5
# Navigate to the new working tree
cd .worktrees/backport-release/3.5
# Create a new branch
git switch --create backport-6244-to-release/3.5
# Cherry-pick the merged commit of this pull request and resolve the conflicts
git cherry-pick d267efd7163932690bf0fd36fff13ea87d3aaff8,aa3e1e39f52228140fdf1d4c64c63ca10d84e641,c24b3c61f8ac56568783ad026233ebaaf34939d9,d0590ccd74e436cf2929f7d2ae262182b6ba5f43
# Push it to GitHub
git push --set-upstream origin backport-6244-to-release/3.5
# Go back to the original working tree
cd ../..
# Delete the working tree
git worktree remove .worktrees/backport-release/3.5
```
Then, create a pull request where the `base` branch is `release/3.5` and the `compare`/`head` branch is `backport-6244-to-release/3.5`.

**Pull request merged**